tests/integ/test_nested_select_plan_analysis.py

```diff
@@ -44,51 +55,161 @@ def setup(request, session):
+    """
     return session.create_dataframe(
         [[1, "a", 2], [2, "b", 3], [3, "c", 7]], schema=["a", "b", "c"]
     )
 
 
+@pytest.fixture(scope="function")
+def sample_table(session):
+    table_name = Utils.random_table_name()
+    Utils.create_table(
+        session, table_name, "a int, b int, c int, d int", is_temporary=True
+    )
+    session._run_query(
+        f"insert into {table_name}(a, b, c, d) values " "(1, 2, 3, 4), (5, 6, 7, 8)"
+    )
+    yield table_name
+    Utils.drop_table(session, table_name)
+
+
 def verify_dataframe_select_statement(
-    df: DataFrame, can_be_merged_when_enabled: bool
+    df: DataFrame,
+    can_be_merged_when_enabled: bool,
+    complexity_before_merge: Dict[PlanNodeCategory, int],
+    complexity_after_merge: Optional[Dict[PlanNodeCategory, int]] = None,
 ) -> None:
     assert isinstance(df._plan.source_plan, SelectStatement)
 
     if not df.session.large_query_breakdown_enabled:
         # if large query breakdown is disabled, _merge_projection_complexity_with_subquery will always be false
         assert df._plan.source_plan._merge_projection_complexity_with_subquery is False
+        assert_df_subtree_query_complexity(df, complexity_before_merge)
     else:
         assert (
             df._plan.source_plan._merge_projection_complexity_with_subquery
             == can_be_merged_when_enabled
         )
+        if can_be_merged_when_enabled:
+            assert (
+                complexity_after_merge
+            ), "no complexity after merge is provided for validation"
+            assert_df_subtree_query_complexity(df, complexity_after_merge)
+        else:
+            assert_df_subtree_query_complexity(df, complexity_before_merge)
 
 
 def test_simple_valid_nested_select(simple_dataframe):
     df_res = simple_dataframe.select((col("a") + 1).as_("a"), "b", "c").select(
         (col("a") + 3).as_("a"), "c"
     )
-    verify_dataframe_select_statement(df_res, can_be_merged_when_enabled=True)
+    verify_dataframe_select_statement(
+        df_res,
+        can_be_merged_when_enabled=True,
+        complexity_before_merge={
+            PlanNodeCategory.LOW_IMPACT: 2,
+            PlanNodeCategory.COLUMN: 8,
+            PlanNodeCategory.LITERAL: 11,
+        },
+        complexity_after_merge={
+            PlanNodeCategory.LOW_IMPACT: 2,
+            PlanNodeCategory.LITERAL: 11,
+            PlanNodeCategory.COLUMN: 5,
+        },
+    )
     # add one more select
     df_res = df_res.select(col("a") * 2, (col("c") + 2).as_("d"))
-    verify_dataframe_select_statement(df_res, can_be_merged_when_enabled=True)
+    verify_dataframe_select_statement(
+        df_res,
+        can_be_merged_when_enabled=True,
+        complexity_before_merge={
+            PlanNodeCategory.LOW_IMPACT: 4,
+            PlanNodeCategory.COLUMN: 10,
+            PlanNodeCategory.LITERAL: 13,
+        },
```

Comment:
this is an interesting pattern. I would have assumed that after merge, the complexity goes up but it is going down in almost all tests I have seen. is this expected?

Reply:
yes. Under some case it will, think about the effect is very similar as SQL simpllifier, when we have SELECT col1 + 1, col2 from (select col1 + 2 as col1, col2 from table), the effect is like select col1 + 1 + 2, col2 from table, you can see that the query after merging is actually simplier.

The case that will start exploding is the case when there are more than one dupliation, like select col1 + 1 + col2, col2 from (select col1 + 2 as col1, col2 + 2  as col2 from table), the final effect is like select (col1 + 2) + 1 + (col2 + 2), col2 + 2 form table, the inlining will make the total complexity increase.

Reply:
if you take a look of the test case like test_large_query_breakdown_with_cte_optimization, i actually added more reference to the col, and the complexity increase much faster, and the number of loops to reach complexity boundary becomes samller